external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+    /// `Indexed` locations.
+    #[derive(Debug, Clone)]
+    enum RuntimeLocation {
+        Stack(usize),
+        Local(TraceIndex, usize),
+        Indexed(Box<RuntimeLocation>, usize),
+        Global(TraceIndex),
+    }
+
+    /// The reference information for a local. This is used to track the state of a local in a frame.
+    /// * It can be a value, in which case the reference type is `Value`.
+    /// * It can be a local that does not currently hold a value (is "empty"), in which case the
+    ///   we track the reference type and the type of the local, but we don't have a `RuntimeLocation`
+    ///   for the reference.
+    /// * It can be a local that holds a value (is "filled"), in which case we track the reference type and the
+    ///   location the reference resolves to.
+    #[derive(Debug, Clone)]
+    enum ReferenceType {
+        Value,
+        Empty {
+            ref_type: RefType,
+        },
+        Filled {
+            ref_type: RefType,
+            location: RuntimeLocation,
+        },
+    }
+
+    /// A `RootedType` is a a type layout with reference information, where any reference type is
+    /// fully rooted back to a specific location.
+    #[derive(Debug, Clone)]
+    struct RootedType {
+        layout: MoveTypeLayout,
+        ref_type: Option<(RefType, RuntimeLocation)>,
+    }
+
+    /// A `LocalType` is a type layot where a reference type may not be rooted to a
+    /// specific location (or it may be rooted to a specific location if the location is filled with a
+    /// value at the time). Not the type layout may be `None` in the case where the type is not
+    /// calculable at runtime without error.
+    #[derive(Debug, Clone)]
+    struct LocalType {
+        layout: Option<MoveTypeLayout>,
+        ref_type: ReferenceType,
+    }
+
+    impl TagWithLayoutInfoOpt {
+        pub fn into_tag_with_refs(&self) -> TypeTagWithRefs {
+            TypeTagWithRefs {
+                type_: self.tag.clone(),
+                ref_type: self.layout.1.clone(),
+            }
+        }
+    }
+
+    impl RuntimeLocation {
+        fn into_trace_location(&self) -> Location {
+            match self {
+                RuntimeLocation::Stack(_) => {
+                    panic!("Cannot convert stack location to trace location")
+                }
+                RuntimeLocation::Local(fidx, lidx) => Location::Local(*fidx, *lidx),
+                RuntimeLocation::Indexed(loc, idx) => {
+                    Location::Indexed(Box::new(loc.into_trace_location()), *idx)
+                }
+                RuntimeLocation::Global(id) => Location::Global(*id),
+            }
+        }
+
+        fn into_runtime_location(loc: Location) -> Self {
+            match loc {
+                Location::Local(fidx, lidx) => RuntimeLocation::Local(fidx, lidx),
+                Location::Indexed(loc, idx) => RuntimeLocation::Indexed(
```

Comment:
Can you have an indexed runtime location that is rooted on the stack?

Reply:
Nope! always eventually into a local or global.

Reply:
Okay, because then my follow-up question was whether it would be simpler to separate out the reference part of the RuntimeLocation from the stack part, so that you can avoid recursively converting from `Location` to `RuntimeLocation`:

```rust
enum RuntimeLocation {
    Stack(u16),
    Frame(Location),
}
```